### ISSUE TITLE
Download plugins in initContainer instead of jenkins-master

### ISSUE THREAD

**Issue by apogrebnyak** [jenkinsci/kubernetes-operator#337]
Currently plugins are downloaded and installed during jenkins-master initialization.

This has a few drawbacks:

1. Time of the download is unpredictable (It depends on connection speed, rate limiting by plugin mirror sites, and the actual list of plugins needed)
1. Timeout and initial delays of liveness and readiness probes cannot be fixed, as they depend on the time needed to download variable components
1. It's hard to tell if jenkins-master is failing because of misconfiguration, or because of a problem with plugins download.

I suggest moving plugins download to `initContainer` which should download them on a shared volume, which jenkins-master can copy to `${REF}/plugins`.

The `initContainer` should be of the same image as `jenkins-master` and will have a stock working copy of `install-plugins.sh`.

**Comment by apogrebnyak** [comment]
This ticket was requested here -> https://github.com/jenkinsci/kubernetes-operator/issues/333#issuecomment-616512021

**Comment by evberrypi** [comment]
Also seeing this issue when adding more than 3 plugins to the Jenkins Operator Helm Chart.

**Comment by akram** [comment]
Hi @apogrebnyak ,
cc @anxolerd , @tomaszsek 

I agree with the forementionned issues, however, I don't see how using init container will solve this issue: Having an init container will only separate the 2 phases to allow easier diagnostic in case of Jenkins failing to startup, but

- it will not prevent Jenkins to not start if donwloads are slow or update center is broken
- it will increase startup time
- if jenkins or one of the init container does not start properly, we will endup in the same restart loop with additional re-download step.

To address this issue, I don't see another way than using immutability principle behind that is inherent of use of container images. It means, that use should bring its own pre-configured image with its set of plugins.

So, basically, it would be better to remove download of plugins at startup time as this prevents immutability.

**Comment by apogrebnyak** [comment]
Hi @akram,

I think you are onto something as having pre-baked image with all plugins needed for Jenkins to operate.

However, there is still a value with `initContainer`.  The issue is really with readiness and liveness checks, which are the properties of a `container` not `pod` as a whole.

Once the variable of downloading dependencies is removed from main Jenkins-CR container startup, the liveness and readiness values can be tightened much more precisely.

**Comment by anxolerd** [comment]
@akram, i am not sure that removing plugins download is a good idea.
Despite i do know the case, where image with preinstalled plugins is used, i consider it quite difficult to maintain (multiple entrypoints, etc) as well as building and maintaining own image may not be always desirable.

Moreover, when plugins are managed separately in the image, image definition becomes decoupled from jobs definition therefore making whole Jenkins configuration span over lots of locations.

**Comment by waveywaves** [comment]
@anxolerd even if the process might be a bit slower, the operator becomes more maintainable and jenkins instances are faster to start. If you have to download plugins everytime a new Jenkins instance needs to be spun up, I don't know how great that would be for production. 

Downloading plugins using initContainer makes sense if
- The Plugins can be downloaded and the Jenkins Servers are up 24/7 with no downtime (which is not the case)
- Plugins are downloaded in a matter of milliseconds

Cons to building an image for Jenkins everytime include
- Have to build a complete image everytime for Jenkins and takes a bit of time
- e2e tests on the new image to see if it is running fine

But if that guarantees if the user will have a seamless experience with Jenkins, the pros overweigh the cons in this one. 

The image needn't be built by the user everytime and the operator can help with that to have a secure yet customizable Jenkins image for the user.

With the image builds we can also see to it that the images can have a different dumb-init process as the parent process so that no issues with zombies ever occur.

To start off, downloading images by default is definitely not a great idea for production if you have to download plugins for every new instance of the CR.

I will add this topic to the agenda so we can discuss this in the next meeting. This definitely needs a discussion to weigh the pros and cons of each one of the things in terms of production usage of Jenkins.

**Comment by apogrebnyak** [comment]
There is another issue with downloading plugins from Jenkins website.

For some environments it may not be allowed because of internal security processes, where all pieces must be vetted and only approved versions installed.

Having ability to use a pre-baked plugins image may address this requirement as well.

**Comment by waveywaves** [comment]
I have added the agenda item for the upcoming 21st May, 2020 meeting for discussion.

**Comment by Fixmetal** [comment]
I stumbled on this issue since I've prepared a Jenkins image with pre-installed plugins but can't make the operator work without defining the operator needed plugins into .spec.master.basePlugins section.
I suppose the discussion has never come to an end...?
